### PR TITLE
Improve TLS client side session cache references

### PR DESCRIPTION
**Pull request by dgarske**
Improve TLS client side session cache references to provide option for not returning an internal session cache pointer. Added reference counter logic to `wolfSSL_get1_sesson` that requires calling `wolfSSL_SESSION_free`. To disable this feature use `NO_SESSION_CACHE_REF`.
ZD13363

### CODE REVIEW

**Comment by dgarske** [comment]
@SparkiDev can you review this? Please be critical and hold off merging until I green light.

File: src/ssl.c [hunk 1]
```diff
@@ -12182,7 +12185,7 @@ WOLFSSL_SESSION* wolfSSL_get_session(WOLFSSL* ssl)
 {
     WOLFSSL_ENTER("SSL_get_session");
     if (ssl)
-        return GetSession(ssl, 0, 1);
+        return GetSession(ssl, NULL, 1);
```
Comment (SparkiDev) [inline src/ssl.c:12188]: OpenSSL has two functions, SSL_get_session() and SSL_get1_session().
SSL_get1_session() returns the session but increased the reference count. Therefore you have to free it,
SSL_get_session() returns the session as a 'reference'. That is, no free needed after.

File: src/ssl.c [hunk 2]
```diff
@@ -23183,18 +23340,32 @@ WOLFSSL_SESSION* wolfSSL_SESSION_dup(WOLFSSL_SESSION* session)
...
-void FreeSession(WOLFSSL_SESSION* session, int isAlloced)
+#endif /* OPENSSL_EXTRA || HAVE_EXT_CACHE */
+
+void FreeSession(WOLFSSL_SESSION* session, void* heap)
 {
     if (session == NULL)
         return;
 
+#ifdef WOLFSSL_STATIC_MEMORY
+    if (heap == NULL)
```
Comment (SparkiDev) [inline src/ssl.c:23357]: Why does the heap need to be passed in?
Doesn't the session have the heap pointer kept against it?

File: src/ssl.c [hunk 3]
```diff
@@ -23169,12 +23326,12 @@ WOLFSSL_SESSION* wolfSSL_SESSION_dup(WOLFSSL_SESSION* session)
         copy->refCount = 1;
 #endif
 #ifdef HAVE_SESSION_TICKET
-        if (session->isDynamic) {
+        if (session->ticketLenAlloc > 0) {
```
Comment (SparkiDev) [inline src/ssl.c:23329]: Ticket is allocated with NULL here and 'heap' in other places.

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: wolfssl/internal.h [hunk 4]
```diff
@@ -3396,23 +3407,55 @@ struct WOLFSSL_SESSION {
...
+#ifdef WOLFSSL_STATIC_MEMORY
+    void*              heap;
+#endif
+
+#ifdef ENABLE_CLIENT_SESSION_REF
+    /* pointer to WOLFSSL_SESSION in internal cache (for WOLFSSL_SESSION_TYPE_REF) */
+    void*              refPtr;
+#endif
+
+    /* below are not allocated for the WOLFSSL_SESSION_TYPE_REF, but pointers to
```
[the remark this answers is not in the document]
Comment (SparkiDev) [reply wolfssl/internal.h]: Makes sense now! Thanks!

File: src/ssl.c (hunk 1 above)
Comment (SparkiDev) [reply src/ssl.c:12188]: We need wolfSSL_get_session() to match the OpenSSL API and not return a session that needs to be freed.

**Review by SparkiDev — APPROVED** [review]

File: src/ssl.c [hunk 5]
```diff
@@ -15243,140 +15277,177 @@ WOLFSSL_SESSION* GetSession(WOLFSSL* ssl, byte* masterSecret,
...
     }
     else {
 #if defined(OPENSSL_EXTRA) && defined(WOLFSSL_ERROR_CODE_OPENSSL)
         WOLFSSL_MSG("Session is expired but return success for \
                               OpenSSL compatibility");
-        return WOLFSSL_SUCCESS;
+        ret = WOLFSSL_SUCCESS;
 #endif /* OPENSSL_EXTRA && WOLFSSL_ERROR_CODE_OPENSSL */
-        return WOLFSSL_FAILURE;  /* session timed out */
+        ret = WOLFSSL_FAILURE;  /* session timed out */
```
Comment (SparkiDev) [inline src/ssl.c:15448]: Previously this line was not invoked when the above lines were compiled in.

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: src/ssl.c (hunk 5 above)
Comment (dgarske) [reply src/ssl.c:15448]: Great eye! Pushed fix.

**Review by SparkiDev — APPROVED** [review]

**Pull request merged**